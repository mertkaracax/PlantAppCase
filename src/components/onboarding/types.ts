import { ReactNode } from "react";
import { DimensionValue } from "react-native";

export type HeaderProps = {
  title: ReactNode;
  subtitle?: string | ReactNode;
  width?: number | "auto" | DimensionValue;
  height?: number | "auto" | DimensionValue;
  top?: number | "auto" | DimensionValue;
  left?: number | "auto" | DimensionValue;
};
