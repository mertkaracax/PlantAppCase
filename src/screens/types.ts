export interface RouteProps {
  navigation?: any;
  route?: any;
}

export enum NavigatorNames {
  OnboardingNavigator = "Onboarding",
  BottomTabNavigator = "BottomTab",
}

export enum OnboardingScreenNames {
  ONBOARDING_ONE = "OnboardingOne",
  ONBOARDING_TWO = "OnboardingTwo",
  ONBOARDING_THREE = "OnboardingThree",
}
