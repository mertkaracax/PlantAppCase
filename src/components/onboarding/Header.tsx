import { StyleSheet, Text, View } from "react-native";
import React from "react";
import { HeaderProps } from "./types";
import { ch, cw } from "@src/style/dimensions";
import { FontSize } from "@src/style/fonts";
import { Color } from "@src/style/colors";
import { FontStyles } from "@src/style/general";

const Header: React.FC<HeaderProps> = ({
  title,
  subtitle,
  width = "auto",
  height = "auto",
}) => {
  return (
    <View style={[styles.headerContainer, { width: width, height: height }]}>
      {title}
      {subtitle && (
        <Text style={[styles.subtitle, FontStyles.rubik]}>{subtitle}</Text>
      )}
    </View>
  );
};

export default Header;

const styles = StyleSheet.create({
  headerContainer: {
    marginLeft: cw(24),
    marginTop: ch(12),
    padding: 8,
    lineHeight: 33.3,
  },
  title: {
    fontWeight: "300",
    fontSize: FontSize.SIZE28,
    color: Color.HeaderText,
  },
  subtitle: {
    color: Color.SubheaderText,
    fontWeight: "300",
    fontSize: FontSize.SIZE16,
  },
});
