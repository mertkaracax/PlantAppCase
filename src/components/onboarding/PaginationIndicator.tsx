import { StyleSheet, Text, View } from "react-native";
import React, { useState } from "react";
import { OnboardingScreenNames } from "@src/screens/types";
import { ch, cw } from "@src/style/dimensions";
import { DotSize } from "@src/style/fonts";

type PaginationProps = {
  active_screen: OnboardingScreenNames;
};

const PaginationIndicator: React.FC<PaginationProps> = ({ active_screen }) => {
  const pages: Array<OnboardingScreenNames> = Object.values(
    OnboardingScreenNames
  );

  const active_index: number = Object.keys(OnboardingScreenNames).indexOf(
    active_screen
  );

  return (
    <View style={styles.container}>
      {pages.map((page) => (
        <View
          style={page === active_screen ? styles.activeDot : styles.dot}
        ></View>
      ))}
    </View>
  );
};

export default PaginationIndicator;

const styles = StyleSheet.create({
  container: {
    width: cw(38),
    height: ch(10),
    position: "absolute",
    top: ch(755.5),
    flexDirection: "row",
    justifyContent: "space-between",
    alignSelf: "center",
    alignItems: "center",
  },
  dot: {
    width: DotSize.SMALL,
    height: DotSize.SMALL,
    backgroundColor: "gray",
    borderRadius: DotSize.SMALL,
  },
  activeDot: {
    width: DotSize.BIG,
    height: DotSize.BIG,
    borderRadius: DotSize.BIG_RADIUS,
    backgroundColor: "black",
  },
});
