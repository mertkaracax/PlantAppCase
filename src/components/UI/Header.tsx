import { StyleSheet, Text, View } from "react-native";
import React, { memo } from "react";
import { HeaderProps } from "../onboarding/types";
import { ch, cw } from "@src/style/dimensions";
import { FontStyles } from "@src/style/general";

const Header: React.FC<HeaderProps> = ({
  title,
  subtitle,
  width = cw(315),
  height = "auto",
  top,
  left,
  marginTop = ch(12),
  marginBottom,
  maxWidth,
  children,
}) => {
  return (
    <View
      style={[
        styles.headerContainer,
        {
          width: width,
          maxWidth: maxWidth,
          height: height,
          top: top,
          left: left,
          marginTop: marginTop,
          marginBottom: marginBottom,
        },
      ]}
    >
      {title}
      {typeof subtitle === "string" ? (
        <Text style={[FontStyles.headerSubtitle]}>{subtitle}</Text>
      ) : (
        React.isValidElement(subtitle) && subtitle
      )}
      {children}
    </View>
  );
};

export default memo(Header);

const styles = StyleSheet.create({
  headerContainer: {
    marginLeft: cw(24),
    marginTop: ch(12),
  },
});
