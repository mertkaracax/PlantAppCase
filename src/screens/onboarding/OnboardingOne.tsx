import { View, Text, StyleSheet, Image, ImageBackground } from "react-native";
import React from "react";
import { SafeAreaView } from "react-native-safe-area-context";
import { OnboardingScreenNames, OnboardingScreenProps } from "../types";
import { Font, LetterSpace, LineHeight } from "@src/style/fonts";
import { FontStyles } from "@src/style/general";
import Header from "@src/components/UI/Header";
import { SCREEN, ch, cw } from "@src/style/dimensions";
import { IMAGES } from "@src/assets/images";
import FlowButton from "@src/components/UI/FlowButton";
import PaginationIndicator from "@src/components/onboarding/PaginationIndicator";
import { useBackBlocker } from "@src/hooks/navigationHooks";

const OnboardingOne: React.FC<OnboardingScreenProps> = ({
  navigation,
  route,
}) => {
  useBackBlocker();
  const headerTitle = (
    <Text
      numberOfLines={2}
      style={[
        FontStyles.headerTitle,
        {
          fontFamily: Font.Medium,
          letterSpacing: LetterSpace.TIGHT,
          lineHeight: LineHeight.LINE33_18,
        },
      ]}
    >
      Take a photo to{" "}
      <Text style={{ fontFamily: Font.ExtraBold }}>identify {`\n`}</Text> the
      plant!
    </Text>
  );

  return (
    <ImageBackground
      style={{ flex: 1 }}
      source={require("@src/assets/images/onboarding/o1.png")}
    >
      <SafeAreaView style={{ flex: 1 }}>
        <Header title={headerTitle} height={ch(75)}>
          <Image source={IMAGES.onboarding_2.brush} style={styles.brush} />
        </Header>

        <View style={styles.imageContainer}>
          <Image style={styles.image} source={IMAGES.onboarding_2.background} />
        </View>
        <FlowButton
          style={{ height: ch(56) }}
          title="Continue"
          onPress={() => {
            navigation.navigate(OnboardingScreenNames.ONBOARDING_TWO);
          }}
        />
        <PaginationIndicator
          active_screen={OnboardingScreenNames.ONBOARDING_ONE}
        />
      </SafeAreaView>
    </ImageBackground>
  );
};

const styles = StyleSheet.create({
  brush: {
    width: cw(138),
    height: ch(12),
    position: "absolute",
    right: cw(0),
    bottom: ch(25.29),
  },
  imageContainer: {
    height: ch(530),
    marginTop: ch(12),
    width: "100%",
  },
  image: {
    position: "absolute",
    width: SCREEN.height * 0.46,
    height: SCREEN.height * 0.81,
    top: ch(10),
    alignSelf: "center",
  },
});

export default OnboardingOne;
