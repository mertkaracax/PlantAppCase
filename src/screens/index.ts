import OnboardingOne from "../screens/onboarding/OnboardingOne";
import OnboardingTwo from "./onboarding/OnboardingTwo";
import OnboardingThree from "./onboarding/OnboardingThree";
import Home from "./home/Home";

export { Home, OnboardingOne, OnboardingTwo, OnboardingThree };
