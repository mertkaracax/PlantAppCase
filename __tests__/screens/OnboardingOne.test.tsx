import React from "react";
import OnboardingOne from "@src/screens/onboarding/OnboardingOne";
import renderer, {
  ReactTestRenderer,
  ReactTestInstance,
} from "react-test-renderer";
import { OnboardingScreenNames } from "@src/screens/types";
import Header from "@src/components/UI/Header";
import PaginationIndicator from "@src/components/onboarding/PaginationIndicator";

jest.mock("@src/components/UI/Header", () => "Header");
jest.mock(
  "@src/components/onboarding/OnboardingButton",
  () => "OnboardingButton"
);
jest.mock(
  "@src/components/onboarding/PaginationIndicator",
  () => "PaginationIndicator"
);
jest.mock("react-native-safe-area-context", () => ({
  SafeAreaView: "SafeAreaView",
}));
jest.mock("@src/hooks/navigationHooks", () => ({
  useBackBlocker: jest.fn(),
}));
jest.mock("@src/assets/images", () => ({
  IMAGES: {
    onboarding_2: {
      brush: "brush.png",
      background: "background.png",
    },
  },
}));

describe("OnboardingOne", () => {
  let component: ReactTestRenderer;
  let root: ReactTestInstance;
  const mockNavigate = jest.fn();

  beforeEach(() => {
    component = renderer.create(
      <OnboardingOne navigation={{ navigate: mockNavigate }} route={{}} />
    );
    root = component.root;
  });

  test("renders correctly", () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("contains correct header title", () => {
    const header = root.findByType(Header);
    expect(header.props.title.props.children).toContain("Take a photo to");
  });

  test("navigates to OnboardingTwo when Continue button is pressed", () => {
    const button = root.findByProps({ title: "Continue" });
    button.props.onPress();
    expect(mockNavigate).toHaveBeenCalledWith(
      OnboardingScreenNames.ONBOARDING_TWO
    );
  });

  test("renders PaginationIndicator with correct active screen", () => {
    const paginationIndicator = root.findByType(PaginationIndicator);
    expect(paginationIndicator.props.active_screen).toBe(
      OnboardingScreenNames.ONBOARDING_ONE
    );
  });

  test("displays brush image", () => {
    const brushImage = root.findByProps({ source: "brush.png" });
    expect(brushImage).toBeTruthy();
  });

  test("displays background image", () => {
    const backgroundImage = root.findByProps({ source: "background.png" });
    expect(backgroundImage).toBeTruthy();
  });
});
