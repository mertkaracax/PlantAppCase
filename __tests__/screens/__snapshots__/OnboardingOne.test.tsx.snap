// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnboardingOne renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    {
      "flex": 1,
    }
  }
>
  <Image
    source={1}
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <SafeAreaView
    style={
      {
        "flex": 1,
      }
    }
  >
    <Header
      height={123.21428571428571}
      title={
        <Text
          numberOfLines={2}
          style={
            [
              {
                "color": "#13231B",
                "fontFamily": "Rubik_400Regular",
                "fontSize": 56.00000000000001,
                "letterSpacing": 0.14,
                "lineHeight": 33.18,
                "maxWidth": 630,
              },
              {
                "fontFamily": "Rubik_500Medium",
                "letterSpacing": -2,
                "lineHeight": 54.510000000000005,
              },
            ]
          }
        >
          Take a photo to
           
          <Text
            style={
              {
                "fontFamily": "Rubik_800ExtraBold",
              }
            }
          >
            identify 
            

          </Text>
           the plant!
        </Text>
      }
    >
      <Image
        source="brush.png"
        style={
          {
            "bottom": 41.54785714285714,
            "height": 19.71428571428571,
            "position": "absolute",
            "right": 0,
            "width": 276,
          }
        }
      />
    </Header>
    <View
      style={
        {
          "height": 870.7142857142858,
          "marginTop": 19.71428571428571,
          "width": "100%",
        }
      }
    >
      <Image
        source="background.png"
        style={
          {
            "alignSelf": "center",
            "height": 1080.54,
            "position": "absolute",
            "top": 16.42857142857143,
            "width": 613.64,
          }
        }
      />
    </View>
    <OnboardingButton
      onPress={[Function]}
      style={
        {
          "height": 92,
        }
      }
      title="Continue"
    />
    <PaginationIndicator
      active_screen="OnboardingOne"
    />
  </SafeAreaView>
</View>
`;
